File: AndroidArchitecture/mvp-rxjava/src/main/java/com/codearms/maoqiqi/app/statistics/StatisticsFragment.java
package com.codearms.maoqiqi.app.statistics;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.codearms.maoqiqi.app.R;
import com.codearms.maoqiqi.app.base.BaseFragment;
import com.codearms.maoqiqi.app.utils.MessageMap;
import com.codearms.maoqiqi.app.utils.SnackbarUtils;

/**
 * Main UI for the statistics screen.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 13:32
 */
public class StatisticsFragment extends BaseFragment implements StatisticsContract.View {

    private StatisticsContract.Presenter presenter;

    private TextView tvStatistics;

    @Override
    public void setPresenter(StatisticsContract.Presenter presenter) {
        this.presenter = presenter;
    }

    @Override
    public boolean isActive() {
        return isAdded();
    }

    public static StatisticsFragment newInstance() {
        return new StatisticsFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View root = inflater.inflate(R.layout.fragment_statistics, container, false);
        tvStatistics = root.findViewById(R.id.tvStatistics);
        return root;
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.subscribe();
    }

    @Override
    public void onPause() {
        super.onPause();
        presenter.unsubscribe();
    }

    @Override
    public void showStatistics(int numberOfActiveTasks, int numberOfCompletedTasks) {
        tvStatistics.setText(getString(R.string.statistics_info, numberOfActiveTasks, numberOfCompletedTasks));
    }

    @Override
    public void showMessage(String message) {
        SnackbarUtils.show(getView(), MessageMap.get(message));
    }
}

File: AndroidArchitecture/mvp-kotlin/src/main/java/com/codearms/maoqiqi/app/statistics/StatisticsFragment.kt
package com.codearms.maoqiqi.app.statistics

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.codearms.maoqiqi.app.R
import com.codearms.maoqiqi.app.base.BaseFragment
import com.codearms.maoqiqi.app.utils.MessageMap
import com.codearms.maoqiqi.app.utils.show

/**
 * Main UI for the statistics screen.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 13:32
 */
class StatisticsFragment : BaseFragment(), StatisticsContract.View {

    override lateinit var presenter: StatisticsContract.Presenter
    override val isActive: Boolean
        get() = isAdded

    private lateinit var tvStatistics: TextView

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val root = inflater.inflate(R.layout.fragment_statistics, container, false)
        tvStatistics = root.findViewById(R.id.tvStatistics)
        return root
    }

    override fun onResume() {
        super.onResume()
        presenter.start()
    }

    override fun showStatistics(numberOfActiveTasks: Int, numberOfCompletedTasks: Int) {
        tvStatistics.text = getString(R.string.statistics_info, numberOfActiveTasks, numberOfCompletedTasks)
    }

    override fun showMessage(message: String) {
        view?.show(MessageMap.get(message))
    }

    companion object {
        fun newInstance(): StatisticsFragment = StatisticsFragment()
    }
}